src/julia_threads.h

```diff
@@ -74,6 +77,17 @@ typedef struct _jl_tls_states_t {
 #  define JL_CPU_WAKE_NOOP 1
 #endif
 
+// Copied from libuv. Add `JL_CONST_FUNC` so that the compiler
+// can optimize this better.
+static inline unsigned long JL_CONST_FUNC jl_thread_self(void)
+{
+#ifdef _OS_WINDOWS_
+    return (unsigned long)GetCurrentThreadId();
+#else
+    return (unsigned long)pthread_self();
+#endif
+}
```

Comment:
Not all of pthreads is implemented in all UNIX-like systems, so unfortunately, it isn't just the Windows thread API and the pthreads API as in this case, but quite a mess of ``#ifdef``s. I'm happy to have our own threading API wrappers if folks agree, but I do think they should be collected in one place rather than having a mess of conditionals in a number of different places.